packages/lending/src/adapter.ts

```diff
@@ -174,78 +174,86 @@ export class Adapter extends common.Web3Toolkit {
           if (protocol.isAssetTokenized(marketId, srcToken)) {
             flashLoanRepay.subWei(2);
           }
           const flashLoanAggregatorQuotation = await apisdk.protocols.utility.getFlashLoanAggregatorQuotation(
             this.chainId,
             { repays: [flashLoanRepay], protocolId: protocol.preferredFlashLoanProtocolId }
           );
-          const [flashLoanLoanLogic, flashLoanRepayLogic] = apisdk.protocols.utility.newFlashLoanAggregatorLogicPair(
-            flashLoanAggregatorQuotation.protocolId,
-            flashLoanAggregatorQuotation.loans.toArray()
-          );
-          output.logics.push(flashLoanLoanLogic);
-
-          // 3. ---------- swap ----------
-          // swap the flashloan borrow amount to dest collateral
+          // 2-3. swap the flashloan borrow amount to dest collateral
           const swapper = this.findSwapper([srcToken.wrapped, destToken.wrapped]);
           const swapQuotation = await swapper.quote({
             input: flashLoanAggregatorQuotation.loans.get(srcToken.wrapped),
             tokenOut: destToken.wrapped,
             slippage,
           });
-          const swapTokenLogic = swapper.newSwapTokenLogic(swapQuotation);
-          output.logics.push(swapTokenLogic);
-
-          // 4. ---------- supply ----------
-          // supply to target collateral
+          // 2-4. supply input is swap output
           const supplyInput = swapQuotation.output;
-          const supplyLogic = protocol.newSupplyLogic({ marketId, input: supplyInput });
-          // 4-1. use BalanceLink to prevent swap slippage
-          supplyLogic.fields.balanceBps = common.BPS_BASE;
-          output.logics.push(supplyLogic);
-          output.afterPortfolio.supply(supplyInput.token, supplyInput.amount);
-          // 4-2. set dest amount
+          // 2-5. set dest amount
           output.destAmount = supplyInput.amount;
+          // 2-6. validate supply cap
+          if (!destCollateral.validateSupplyCap(supplyInput.amount)) {
+            output.error = { name: 'destAmount', code: 'SUPPLY_CAP_EXCEEDED' };
+          }
+          output.afterPortfolio.supply(supplyInput.token, supplyInput.amount);
 
-          // 5. ---------- withdraw ----------
-          const withdrawOutput = new common.TokenAmount(srcToken.wrapped, srcAmount);
-          const withdrawLogic = protocol.newWithdrawLogic({ marketId, output: withdrawOutput });
-          // 5-1. if protocol is collateral tokenized
-          if (protocol.isAssetTokenized(marketId, withdrawOutput.token)) {
-            // 5-1-1. return dest protocol token to user
-            const returnFundsLogic = apisdk.protocols.utility.newSendTokenLogic({
-              input: supplyLogic.fields.output!,
-              recipient: account,
-            });
-            returnFundsLogic.fields.balanceBps = common.BPS_BASE;
-            output.logics.push(returnFundsLogic);
+          if (!output.error) {
```

Comment:
wondering if we can reduce the indents in this clause in two ways

```
1.  if(output.error) return ...
2. return after output.error = ....
```

Reply:
Sure, "early return" is one way of writing code, but it might result in many instances of return output. This approach might make the code feel less cohesive when read from top to bottom. What do other people think? @zodahu @LuPoYi

Reply:
Both approaches work for me. However, if I must choose, I have a slight preference for the "early return" approach in this situation.

Reply:
I don't have a strong opinion regarding the adapter.ts case, but we can maintain the current approach since this PR preserves the original style. If you have a strong preference, feel free to create a ticket with your reasons and assign it to me. I'll then allocate resources for evaluation and potential refactoring.